crates/services/gas_price_service/src/v0/service.rs

```diff
@@ -0,0 +1,326 @@
+            .get_metadata(&l2_block_height.into())
+            .map_err(|err| GasPriceError::CouldNotInitUpdater(anyhow::anyhow!(err)))?
+        {
+            algorithm_updater = match old_metadata {
+                UpdaterMetadata::V0(old) => AlgorithmUpdaterV0::new(
+                    old.new_exec_price,
+                    min_exec_gas_price,
+                    exec_gas_price_change_percent,
+                    old.l2_block_height,
+                    l2_block_fullness_threshold_percent,
+                ),
+            };
+        } else {
+            algorithm_updater = AlgorithmUpdaterV0::new(
+                new_exec_price,
+                min_exec_gas_price,
+                exec_gas_price_change_percent,
+                l2_block_height,
+                l2_block_fullness_threshold_percent,
+            );
+        }
+
+        let shared_algo =
+            SharedGasPriceAlgo::new_with_algorithm(algorithm_updater.algorithm());
+        Ok(Self {
+            shared_algo,
+            l2_block_source,
+            metadata_storage,
+            algorithm_updater,
+        })
+    }
+
+    pub fn algorithm_updater(&self) -> &AlgorithmUpdaterV0 {
+        &self.algorithm_updater
+    }
+
+    pub fn next_block_algorithm(&self) -> SharedV0Algorithm {
+        self.shared_algo.clone()
+    }
+
+    async fn update(&mut self, new_algorithm: AlgorithmV0) {
+        self.shared_algo.update(new_algorithm).await;
+    }
+
+    fn validate_block_gas_capacity(
+        &self,
+        block_gas_capacity: u64,
+    ) -> anyhow::Result<NonZeroU64> {
+        NonZeroU64::new(block_gas_capacity)
+            .ok_or_else(|| anyhow!("Block gas capacity must be non-zero"))
+    }
+
+    async fn set_metadata(&mut self) -> anyhow::Result<()> {
+        let metadata: UpdaterMetadata = self.algorithm_updater.clone().into();
+        self.metadata_storage
+            .set_metadata(&metadata)
+            .map_err(|err| anyhow!(err))
+    }
+
+    async fn handle_normal_block(
+        &mut self,
+        height: u32,
+        gas_used: u64,
+        block_gas_capacity: u64,
+    ) -> anyhow::Result<()> {
+        let capacity = self.validate_block_gas_capacity(block_gas_capacity)?;
+
+        self.algorithm_updater
+            .update_l2_block_data(height, gas_used, capacity)?;
+
+        self.set_metadata().await?;
+        Ok(())
+    }
+
+    async fn apply_block_info_to_gas_algorithm(
+        &mut self,
+        l2_block: BlockInfo,
+    ) -> anyhow::Result<()> {
+        match l2_block {
+            BlockInfo::GenesisBlock => {
+                self.set_metadata().await?;
+            }
+            BlockInfo::Block {
+                height,
+                gas_used,
+                block_gas_capacity,
+            } => {
+                self.handle_normal_block(height, gas_used, block_gas_capacity)
+                    .await?;
+            }
+        }
+
+        self.update(self.algorithm_updater.algorithm()).await;
```

Comment:
for reviewers: mostly copied as-is except for an update on line 155 to update the algorithm.